lib/elixir/lib/record.ex

```diff
@@ -85,14 +85,20 @@ defmodule Record do
     case Macro.Env.in_guard?(__CALLER__) do
       true ->
         quote do
-          is_tuple(unquote(data)) and tuple_size(unquote(data)) > 0
-            and :erlang.element(1, unquote(data)) == unquote(kind)
+          :erlang.is_tuple(unquote(data)) and :erlang.tuple_size(unquote(data)) > 0 and
+            :erlang.is_atom(:erlang.element(1, unquote(data))) and
+            :erlang.element(1, unquote(data)) == unquote(kind)
         end
       false ->
         quote do
           result = unquote(data)
-          is_tuple(result) and tuple_size(result) > 0
-            and :erlang.element(1, result) == unquote(kind)
+          cond do
+            :erlang.is_tuple(result) and :erlang.tuple_size(result) > 0 ->
```

Comment:
We don't need to convert it into a cond, do `kind = unquote(kind)` before and check if kind is an atom.